presto-jdbc/src/main/java/com/facebook/presto/jdbc/WarningsManager.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.jdbc;
+
+import com.facebook.presto.spi.PrestoWarning;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.sql.SQLWarning;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class WarningsManager
+{
+    @GuardedBy("this")
+    private int warningsSeen;
+
+    @GuardedBy("this")
+    private SQLWarning firstWarning;
+
+    // To avoid O(n) behavior of SQLException.setNextException keep a reference to the last warning added:
+    @GuardedBy("this")
+    private SQLWarning lastWarning;
+
+    private synchronized void addWarning(PrestoWarning warning)
+    {
+        requireNonNull(warning, "warning is null");
+        if (lastWarning == null) {
+            lastWarning = new PrestoSqlWarning(warning);
+        }
+        else {
+            lastWarning.setNextWarning(new PrestoSqlWarning(warning));
+        }
+        if (firstWarning == null) {
+            firstWarning = lastWarning;
+        }
+        else {
+            lastWarning = lastWarning.getNextWarning();
+        }
+    }
+
+    public synchronized void addWarnings(List<PrestoWarning> warnings)
```

Comment:
This method and logic seems to assume that the list that's being passed in meets the following criteria:
* The list is never smaller than previous invocations
* The new warnings are at the end of list

This is error-prone since it's hard to reason about that API and there's no way to protect against callers using it incorrectly. It's also unclear how this interacts with the `clearWarnings` method below.

The fundamental issue is that this class keeps state (`warningsSeen`) associated with behavior that occurs outside of it. It'd be much cleaner to make it stateless and require callers to pass only the warnings that are new each time.

Also, there may be bug lurking here. I don't think the list obtained from results.getWarnings() in the resultset loop satisfies the above constraints.

Reply:
Makes sense. Changing to Set<WarningCode> to guarantee that only new WarningCodes are added.